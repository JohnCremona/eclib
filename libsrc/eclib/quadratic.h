// quadratic.h: declaration of class for handling integer quadratics
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 

// Stored as bigint* arrays q of size 3 representing q[0]*x^2+q[1]*x+q[2]

#ifndef _ECLIB_QUADRATIC_H
#define _ECLIB_QUADRATIC_H      1 //flags that this file has been included

#include "unimod.h"

class quadratic {
  friend class unimod;
private:
  vector<bigint> coeffs;
public:
  void set(const  bigint& a, const bigint& b, const bigint& c) {coeffs = {a, b, c};}
  quadratic() { bigint zero(0); coeffs={zero,zero,zero};}
  quadratic(const  bigint& a, const bigint& b, const bigint& c) {coeffs = {a, b, c};}
  quadratic(long a, long b, long c)  {coeffs = {BIGINT(a), BIGINT(b), BIGINT(c)};}
  quadratic(const  bigint* q) {coeffs = {q[0], q[1], q[2]};}
  quadratic(const  quadratic& q) {coeffs = q.coeffs;}
  void operator=(const quadratic& q)  {coeffs = q.coeffs;}
  bigint eval(const bigint& x, const bigint& z) const
    {return coeffs[0]*sqr(x) + coeffs[1]*x*z + coeffs[2]*sqr(z);}
  bigint operator()(const bigint& x, const bigint& z) const
    {return coeffs[0]*sqr(x) + coeffs[1]*x*z + coeffs[2]*sqr(z);}
  bigint eval(const bigint& x) const
    {return coeffs[0]*sqr(x) + coeffs[1]*x + coeffs[2];}
  bigint operator()(const bigint& x) const
    {return coeffs[0]*sqr(x) + coeffs[1]*x + coeffs[2];}
  bigint coeff(int i) const {return coeffs[i];}
  bigint operator[](int i) const  {return coeffs[i];}
  void set_coeff(int i, const bigint& a) {if((i>=0)&&(i<=2)) coeffs[i]=a;}
  friend bigint resultant(const quadratic& q1, const quadratic& q2);
  bigint disc() const  {return sqr(coeffs[1])-4*coeffs[0]*coeffs[2];}
  void output(ostream& os=cout) const
    {
      os<<"["<<coeffs[0]<<","<<coeffs[1]<<","<<coeffs[2]<<"]";
    }
  friend ostream& operator<<(ostream& os, const quadratic& q);
  void transform(const unimod& m);
  // In the next 4 functions, m already holds a unimod and is updated:
  void x_shift(const bigint& a, unimod& m);
  void y_shift(const bigint& a, unimod& m);
  void invert(unimod& m);
  void reduce(unimod& m);
};

bigint resultant(const quadratic& q1, const quadratic& q2);

inline ostream& operator<<(ostream& os, const quadratic& q)
{
  return os<<"["<<q.coeffs[0]<<","<<q.coeffs[1]<<","<<q.coeffs[2]<<"]";
}

#endif
