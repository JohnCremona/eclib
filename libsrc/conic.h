// conic.h: declarations of functions for solving conics (see also legendre.h)
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2005 John Cremona
// 
// This file is part of the mwrank package.
// 
// mwrank is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// mwrank is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with mwrank; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
// 
//////////////////////////////////////////////////////////////////////////
 
#ifndef _CONIC_H
#define _CONIC_H      1  //flags that this file has been included

#include "quadratic.h"

bigint cancel1(bigint& x, bigint& y, bigint& z);
     // cancels common factors only, return gcd

void cancel(bigint& x, bigint& y, bigint& z);
     // cancels common factors and leaves z>=0 or z=0 and x>=0

int solve_conic(const bigint& a, const bigint& b, const bigint& c, const bigint& d,
		const vector<bigint>& factorbase,
                bigint& x, bigint& y, bigint& z, int method=4);
int solve_conic(const bigint& a, const bigint& b, const bigint& c, const bigint& d,
                bigint& x, bigint& y, bigint& z, int method=4);
     // Solves axx+bxz+czz=dyy for (x,y,z) not (0,0,0) and returns 1
     // or returns 0 if not possible
     // Should have a, c, d, bb-4ac non-zero
  
int solve_conic(const quadratic& q, const bigint& d,
		bigint& x, bigint& y, bigint& z, int method=4);

int solve_conic(const quadratic& q, const bigint& d,
		       const vector<bigint>& factorbase,
		bigint& x, bigint& y, bigint& z, int method=4);

int solve_conic_diag(const bigint& a, const vector<bigint>& aplist,
		     const bigint& b, const vector<bigint>& bplist,
                bigint& x, bigint& y, bigint& z, int method);
     // Solves xx-azz=byy for (x,y,z) not (0,0,0) and returns 1
     // or returns 0 if not possible
     // Should have a, b non-zero square-free, their prime divisors in aplist, bplist 

void conic_mordell_reduce(const bigint& a, const bigint& b, const bigint& c, bigint& x0, bigint& y0, bigint& z0, int verb=0);
     // Given a>0, b>0, c<0, abc square-free and ax^2+by^2+cz^2=0
     // reduces x, y, z in place using Mordell's method (page 48)
     // to achieve Holzer's bounds |z|<=sqrt(ab) etc.

void conic_diag_reduce(const bigint& a, const bigint& b, bigint& x, bigint& y, bigint& z, int verb=0);
     // As above but with a,b square-free only, calls conic_mordell_reduce

int solve_conic_param(const bigint& a, const bigint& b, const bigint& c, const bigint& d,
		const vector<bigint>& factorbase,
                quadratic& qx, quadratic& qy, quadratic& qz, int method=4, int verb=0);

int solve_conic_param(const bigint& a, const bigint& b, const bigint& c, const bigint& d,
                quadratic& qx, quadratic& qy, quadratic& qz, int method=4, int verb=0);
     // Solves axx+bxz+czz=dyy for (x,y,z) not (0,0,0) and returns 1
     // or returns 0 if not possible
     // Should have a, c, d, bb-4ac non-zero
     // qx,qy,qz are arrays of coeffs of parametrizing quadratics
     //   with leading coeffs qx[0],qy[0],qz[0] one solution

int solve_conic_param(const quadratic& q, const bigint& d,
		      const vector<bigint>& factorbase,
		      quadratic& qx, quadratic& qy, quadratic& qz, 
		      int method=4, int verb=0);

int solve_conic_param(const quadratic& q, const bigint& d,
		      quadratic& qx, quadratic& qy, quadratic& qz, 
		      int method=4, int verb=0);

int testsol(const bigint& a, const bigint& b, const bigint& c, const bigint& d,
                const bigint& x, const bigint& y, const bigint& z, int verb=0);

// Tests to see if a given solution is a non-trivial solution
int testsol(const quadratic& q, const bigint& d,
	    const bigint& x, const bigint& y, const bigint& z, 
	    int verb=0);

int testlocsol(const bigint& a, 
	       const bigint& b, 
	       const bigint& c);
// tests if ax^2+by^2+cz^2=0 is soluble, where a, b, c are pairwise
// coprime and square-free

int testlocsol(const bigint& a, const vector<bigint>& alist, 
	       const bigint& b, const vector<bigint>& blist, 
	       const bigint& c, const vector<bigint>& clist);
// tests if ax^2+by^2+cz^2=0 is soluble, where a, b, c are pairwise
// coprime and square-free, their prime factors being in alist etc.

int testlocsol(const bigint& a, const vector<bigint>& alist, 
	       const bigint& b, const vector<bigint>& blist);
// tests if ax^2+by^2=z^2 is soluble, where a, b are
// square-free, their prime factors being in alist and blist.

int testparamsol(const bigint& a, const bigint& b, const bigint& c, const bigint& d,
                const quadratic& qx, const quadratic& qy, const quadratic& qz, int verb=0);

// Tests to see if a given parametrization is a solution
int testparamsol(const quadratic& q, const bigint& d,
		 const quadratic& qx, const quadratic& qy, const quadratic& qz, 
		 int verb=0);

//miscellaneous test functions:

void testmodsqrt();
void testsqf();
void testcancel();

// Output utilities:

void show_xyz(const bigint& x, const bigint& y, const bigint& z);
void show_cert(const bigint& p, const bigint& q, const bigint& r);
void show_eqn(const bigint& a, const bigint& b, const bigint& c);
void show_eqn_cert(const bigint& a, const bigint& b, const bigint& c,
		   const bigint& p, const bigint& q, const bigint& r);
void show_all(const bigint& a, const bigint& b, const bigint& c, 
	      const bigint& p, const bigint& q, const bigint& r, 
	      bigint& x, bigint& y, bigint& z);

#endif
