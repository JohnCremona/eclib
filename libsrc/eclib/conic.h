// conic.h: declarations of functions for solving conics (see also legendre.h)
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
#ifndef _ECLIB_CONIC_H
#define _ECLIB_CONIC_H      1  //flags that this file has been included

#include "quadratic.h"

ZZ cancel1(ZZ& x, ZZ& y, ZZ& z);
     // cancels common factors only, return gcd

void cancel(ZZ& x, ZZ& y, ZZ& z);
     // cancels common factors and leaves z>=0 or z=0 and x>=0

int solve_conic(const ZZ& a, const ZZ& b, const ZZ& c, const ZZ& d,
		const vector<ZZ>& factorbase,
                ZZ& x, ZZ& y, ZZ& z, int method=4);
int solve_conic(const ZZ& a, const ZZ& b, const ZZ& c, const ZZ& d,
                ZZ& x, ZZ& y, ZZ& z, int method=4);
     // Solves axx+bxz+czz=dyy for (x,y,z) not (0,0,0) and returns 1
     // or returns 0 if not possible
     // Should have a, c, d, bb-4ac non-zero
  
int solve_conic(const quadratic& q, const ZZ& d,
		ZZ& x, ZZ& y, ZZ& z, int method=4);

int solve_conic(const quadratic& q, const ZZ& d,
		       const vector<ZZ>& factorbase,
		ZZ& x, ZZ& y, ZZ& z, int method=4);

int solve_conic_diag(const ZZ& a, const vector<ZZ>& aplist,
		     const ZZ& b, const vector<ZZ>& bplist,
                ZZ& x, ZZ& y, ZZ& z, int method);
     // Solves xx-azz=byy for (x,y,z) not (0,0,0) and returns 1
     // or returns 0 if not possible
     // Should have a, b non-zero square-free, their prime divisors in aplist, bplist 

void conic_mordell_reduce(const ZZ& a, const ZZ& b, const ZZ& c, ZZ& x0, ZZ& y0, ZZ& z0, int verb=0);
     // Given a>0, b>0, c<0, abc square-free and ax^2+by^2+cz^2=0
     // reduces x, y, z in place using Mordell's method (page 48)
     // to achieve Holzer's bounds |z|<=sqrt(ab) etc.

void conic_diag_reduce(const ZZ& a, const ZZ& b, ZZ& x, ZZ& y, ZZ& z, int verb=0);
     // As above but with a,b square-free only, calls conic_mordell_reduce

int solve_conic_param(const ZZ& a, const ZZ& b, const ZZ& c, const ZZ& d,
		const vector<ZZ>& factorbase,
                quadratic& qx, quadratic& qy, quadratic& qz, int method=4, int verb=0);

int solve_conic_param(const ZZ& a, const ZZ& b, const ZZ& c, const ZZ& d,
                quadratic& qx, quadratic& qy, quadratic& qz, int method=4, int verb=0);
     // Solves axx+bxz+czz=dyy for (x,y,z) not (0,0,0) and returns 1
     // or returns 0 if not possible
     // Should have a, c, d, bb-4ac non-zero
     // qx,qy,qz are arrays of coeffs of parametrizing quadratics
     //   with leading coeffs qx[0],qy[0],qz[0] one solution

int solve_conic_param(const quadratic& q, const ZZ& d,
		      const vector<ZZ>& factorbase,
		      quadratic& qx, quadratic& qy, quadratic& qz, 
		      int method=4, int verb=0);

int solve_conic_param(const quadratic& q, const ZZ& d,
		      quadratic& qx, quadratic& qy, quadratic& qz, 
		      int method=4, int verb=0);

int testsol(const ZZ& a, const ZZ& b, const ZZ& c, const ZZ& d,
                const ZZ& x, const ZZ& y, const ZZ& z, int verb=0);

// Tests to see if a given solution is a non-trivial solution
int testsol(const quadratic& q, const ZZ& d,
	    const ZZ& x, const ZZ& y, const ZZ& z, 
	    int verb=0);

int testlocsol(const ZZ& a, 
	       const ZZ& b, 
	       const ZZ& c);
// tests if ax^2+by^2+cz^2=0 is soluble, where a, b, c are pairwise
// coprime and square-free

int testlocsol(const ZZ& a, const vector<ZZ>& alist, 
	       const ZZ& b, const vector<ZZ>& blist, 
	       const ZZ& c, const vector<ZZ>& clist);
// tests if ax^2+by^2+cz^2=0 is soluble, where a, b, c are pairwise
// coprime and square-free, their prime factors being in alist etc.

int testlocsol(const ZZ& a, const vector<ZZ>& alist, 
	       const ZZ& b, const vector<ZZ>& blist);
// tests if ax^2+by^2=z^2 is soluble, where a, b are
// square-free, their prime factors being in alist and blist.

int testparamsol(const ZZ& a, const ZZ& b, const ZZ& c, const ZZ& d,
                const quadratic& qx, const quadratic& qy, const quadratic& qz, int verb=0);

// Tests to see if a given parametrization is a solution
int testparamsol(const quadratic& q, const ZZ& d,
		 const quadratic& qx, const quadratic& qy, const quadratic& qz, 
		 int verb=0);

//miscellaneous test functions:

void testmodsqrt();
void testsqf();
void testcancel();

// Output utilities:

void show_xyz(const ZZ& x, const ZZ& y, const ZZ& z);
void show_cert(const ZZ& p, const ZZ& q, const ZZ& r);
void show_eqn(const ZZ& a, const ZZ& b, const ZZ& c);
void show_eqn_cert(const ZZ& a, const ZZ& b, const ZZ& c,
		   const ZZ& p, const ZZ& q, const ZZ& r);
void show_all(const ZZ& a, const ZZ& b, const ZZ& c,
	      const ZZ& p, const ZZ& q, const ZZ& r,
	      const ZZ& x, const ZZ& y, const ZZ& z);

#endif
