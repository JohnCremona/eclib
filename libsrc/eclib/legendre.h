// legendre.h: declarations of functions for solving legendre equations
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
#ifndef _ECLIB_LEGENDRE_H
#define _ECLIB_LEGENDRE_H      1 //flags that this file has been included

#include "quadratic.h"

void minv(const ZZ& a1, const ZZ& a2, 
	  const ZZ& b1, const ZZ& b2, 
	  const ZZ& c1, const ZZ& c2, 
	  ZZ& xmin, const ZZ& ymin);

// (e+f*i) is (a+b*i) mod (c+d*i) in Z[i]
void GIreduce(const ZZ& a, const ZZ& b, const ZZ& c, 
	      const ZZ& d, ZZ& e, ZZ& f);
// (e+f*i) = gcd ( (a+b*i) , (c+d*i) ) in Z[i]
void GIgcd(const ZZ& a, const ZZ& b, const ZZ& c, 
	   const ZZ& d, ZZ& e, ZZ& f);

// Solve Legendre's equation ax^2+by^2+cz^2=0 using Rusin's reduction, 
// without assuming a,b,c pairwise coprime
// returns 0 if not soluble
int legendre_solve(const ZZ& a, const ZZ& b, const ZZ& c, 
		    ZZ& x, ZZ& y, ZZ& z,
		    int use_lll=0);

int legendre_solve(const ZZ& a, const ZZ& b, const ZZ& c, 
		   const vector<ZZ>& factorbase,
		    ZZ& x, ZZ& y, ZZ& z,
		    int use_lll=0);

// Solve Legendre's equation ax^2+by^2+cz^2=0 using Rusin's reduction, 
// given "certificate" (n,p,q) satisfying a|n^2+bc, b|p^2+ac, c|q^2+ab.
// assumes a,b,c pairwise coprime
void legendre_solve_cert(const ZZ& a, const ZZ& b, const ZZ& c, 
		    const ZZ& n, const ZZ& p, const ZZ& q, 
		    ZZ& x, ZZ& y, ZZ& z);

//Ensure that input is valid
int checkin(const ZZ& a,const ZZ& b,const ZZ& c,
	    const ZZ& n,const ZZ& p,const ZZ& q);

// Check that purported solution is OK
int check_leg(const ZZ& a, const ZZ& b, const ZZ& c,
	      const ZZ& x, const ZZ& y, const ZZ& z);

// Check that purported solution is OK & in correct lattice
int check_leg(const ZZ& a, const ZZ& b, const ZZ& c,
	      const ZZ& n, const ZZ& p, const ZZ& q, 
	      const ZZ& x, const ZZ& y, const ZZ& z);

//Peel off a known square factor  u  from coefficient  a
void lem2a(const ZZ& a, const ZZ& b, const ZZ& c, 
	   const ZZ& n, const ZZ& p, const ZZ& q, 
	   const ZZ& u,
	   ZZ& x, ZZ& y, ZZ& z);

//Peel off a known square factor  u  from coefficient  b
void lem2b(const ZZ& a, const ZZ& b, const ZZ& c, 
	   const ZZ& n, const ZZ& p, const ZZ& q, 
	   const ZZ& u,
	   ZZ& x, ZZ& y, ZZ& z);

//Peel off a known square factor  u  from coefficient  c
void lem2c(const ZZ& a, const ZZ& b, const ZZ& c, 
	   const ZZ& n, const ZZ& p, const ZZ& q, 
	   const ZZ& u,
	   ZZ& x, ZZ& y, ZZ& z);


void lem4(const ZZ& a, const ZZ& b, const ZZ& c, 
	  const ZZ& n, const ZZ& p, const ZZ& q, 
	  ZZ& x, ZZ& y, ZZ& z);

// Finds a certificate or returns 0 if none exists:
int make_certificate(const ZZ& a, const ZZ& b, const ZZ& c, 
		     ZZ& n, ZZ& p, ZZ& q);
int make_certificate(const ZZ& a, const vector<ZZ>& apdivs, 
		     const ZZ& b, const vector<ZZ>& bpdivs, 
		     const ZZ& c, const vector<ZZ>& cpdivs, 
		     ZZ& n, ZZ& p, ZZ& q);

// Check to see if  b is congruent to  +- c  mod a  (assumed positive!)
//  if not, how much to divide  a  by to ensure that congruence holds?
ZZ should(const ZZ& a, const ZZ& b, const ZZ& c);

void legendre_reduce(const ZZ& a, const ZZ& b, const ZZ& c, 
		     ZZ& x0, ZZ& y0, ZZ& z0, int verb=0);
     // Given a, b, c,  ax^2+by^2+cz^2=0
     // reduces x, y, z in place using Mordell's method (page 48)
     // to achieve Holzer's bounds |z|<=sqrt(ab) etc.
  // (just permutes & passes to conic_mordell_reduce())

void new_legendre_reduce(const ZZ& a, const ZZ& b, const ZZ& c, 
		     ZZ& x0, ZZ& y0, ZZ& z0, int verb=0);
     // Given a, b, c,  ax^2+by^2+cz^2=0
     // reduces x, y, z in place using quadratics

void legendre_via_lll(const ZZ& a, const ZZ& b, const ZZ& c, 
		      const ZZ& k1, const ZZ& k2, const ZZ& k3, 
		      ZZ& x, ZZ& y, ZZ& z);

//
// Given one solution, returns quadratics parametrizing all solutions,
// with discriminants -4bc, -4ac, -4ab.  Here a, b, c are assumed
// pairwise coprime but not square-free, and a>0, b>0, c<0.

void legendre_param(const ZZ& a, const ZZ& b, const ZZ& c, 
		    const ZZ& x0, const ZZ& y0, const ZZ& z0, 
		    quadratic& qx, quadratic& qy, quadratic& qz);

//These versions EITHER return 0 with a solution in x,y,z in the lattice
//               OR return 1, 2, 3 and a square factor u of a, b, c (resp)
//               OR return -1 if something fails (should not happen)
int legendre_solve_cert_1(const ZZ& a, const ZZ& b, const ZZ& c, 
			  const ZZ& n, const ZZ& p, const ZZ& q, 
			  ZZ& x, ZZ& y, ZZ& z,
			  ZZ& u);
int lem4_1(const ZZ& a, const ZZ& b, const ZZ& c, 
	   const ZZ& n, const ZZ& p, const ZZ& q, 
	   ZZ& x, ZZ& y, ZZ& z,
	   ZZ& u);

bigfloat holzer_measure(const ZZ& a, const ZZ& b, const ZZ& c, 
			const ZZ& x, const ZZ& y, const ZZ& z);
// max{|a|x^2,|b|y^2,|c|z^2}/|abc|   ( < 1 for a Holzer-reduced solution)

#endif
