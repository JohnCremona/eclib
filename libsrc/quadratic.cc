// quadratic.cc: implementation of class for handling integer quadratics
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
#include "eclib/quadratic.h"

void quadratic::transform(const unimod& m)
{
  ZZ newq0 = eval(m.m11, m.m21);
  ZZ newq2 = eval(m.m12, m.m22);
  coeffs[1] = 2 * (coeffs[0]*m.m11*m.m12 + coeffs[2]*m.m21*m.m22) 
               + coeffs[1] * (m.m11*m.m22+m.m12*m.m21);
  coeffs[0] = newq0;
  coeffs[2] = newq2;
}

void quadratic::x_shift(const ZZ& a, unimod& m)
{
  const ZZ& aq0=a*coeffs[0];
  coeffs[2]+= (aq0+coeffs[1])*a;
  coeffs[1]+= 2*aq0;
  m.x_shift(a);
}

void quadratic::y_shift(const ZZ& a, unimod& m)
{
  const ZZ& aq2=a*coeffs[2];
  coeffs[0]+= (aq2+coeffs[1])*a;
  coeffs[1]+= 2*aq2;
  m.y_shift(a);
}

void quadratic::invert(unimod& m)
{
  swap(coeffs[0],coeffs[2]);  coeffs[1]=-coeffs[1];
  m.invert();
}

void quadratic::reduce(unimod& m)
{
  //  cout<<"Reducing quadratic  "<<(*this)<<endl;
  if(coeffs[0]<0) 
    {
      coeffs[0]=-coeffs[0]; 
      coeffs[2]=-coeffs[2]; 
      coeffs[1]=-coeffs[1];
    }
  ZZ a = roundover(-coeffs[1],2*coeffs[0]);
  x_shift(a,m);
  int reduced = (coeffs[0]<=coeffs[2]);
  while(!reduced)
    {
      invert(m);
      a = roundover(-coeffs[1],2*coeffs[0]);
      x_shift(a,m);
      reduced = (coeffs[0]<=coeffs[2]);
    }
  //  cout<<"Reduced quadratic = "<<(*this)<<endl;
  //  cout<<"  via matrix "<<m<<endl;
  //  cout<<"  of determinant "<<m.det()<<endl;
}

ZZ resultant(const quadratic& q1, const quadratic& q2)
{
  const ZZ&
    a1=q1.coeffs[0], b1=q1.coeffs[1], c1=q1.coeffs[2],
    a2=q2.coeffs[0], b2=q2.coeffs[1], c2=q2.coeffs[2];
  return
    sqr(c2*a1) -a1*c2*b2*b1 + (-2*c2*a2 + sqr(b2))*c1*a1 + c2*a2*sqr(b1) - b2*a2*c1*b1 + sqr(a2*c1);
}
