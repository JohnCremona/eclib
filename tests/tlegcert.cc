// tlegcert.cc: test program for solving legendre equations with certificates
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2005 John Cremona
// 
// This file is part of the mwrank package.
// 
// mwrank is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// mwrank is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with mwrank; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
// 
//////////////////////////////////////////////////////////////////////////
 
#include "marith.h"
#include "quadratic.h"
#include "conic.h"
#include "legendre.h"

#ifndef CONIC_METHOD
#define CONIC_METHOD 4
#endif
//#define TEST_PARAM
//#define HOLZER_MEASURES

int main()
{
  initprimes(string("PRIMES").c_str());
  cout<<"Solving ax^2 + by^2 + cz^2 = 0\n";
  cout<<"Using method "<<CONIC_METHOD<<endl<<endl;

  bigint a,b,c,x,y,z;
  bigint u,k1,k2,k3;
  quadratic qx, qy, qz;

  while(1) 
    {
      cout << "Enter coefficients a b c: ";
      cin >> a >> b >> c;
      cout<<a<<" "<<b<<" "<<c<<endl; 
      if(a==0) abort();
      cout << "Enter certificate k1 k2 k3: ";
      cin >> k1 >> k2 >> k3;
      int use_lll=(CONIC_METHOD==5);
      int res=!use_lll;
      if(use_lll) legendre_via_lll(a,b,c,k1,k2,k3,x,y,z);
      else        res = legendre_solve_cert_1(a,b,c,k1,k2,k3,x,y,z,u);
      if(res)
//    if(!solve_conic(a,0,c,-b,x,y,z,CONIC_METHOD))
	{
	  cout<<"No solution!\n";
	}
      else
	{
#ifdef HOLZER_MEASURES
	  cout<<"Before reduction of solution "; 
	  show_xyz(x,y,z); cout<<endl;
	  cout<<"Holzer measure = "<<holzer_measure(a,b,c,x,y,z)<<endl;
	  cancel1(x,y,z);
	  new_legendre_reduce(a,b,c,x,y,z,0);
	  cout<<"After reduction: ";show_xyz(x,y,z);cout<<endl;
	  cout<<"Holzer measure = "<<holzer_measure(a,b,c,x,y,z)<<endl;
#else
	  cout << "Solution: "; show_xyz(x,y,z); cout<<endl;
#endif // HOLZER_MEASURES
	  

      if(check_leg(a,b,c,x,y,z))
	{
	  cout<<" --OK\n";
#ifdef TEST_PARAM
	  legendre_param(a,b,c,x,y,z,qx,qy,qz);
	  //	  cout<<"Parametric solution:\n";
	  cout<<"x = "<<qx<<" * [u^2,uv,v^2]\n";
	  cout<<"y = "<<qy<<" * [u^2,uv,v^2]\n";
	  cout<<"z = "<<qz<<" * [u^2,uv,v^2]\n";
	  if(testparamsol(a,0,c,-b,qx,qy,qz,0))
	    cout<<"Parametric solution is OK\n";
	  else 
	    cout<<"Parametric solution is wrong!\n";
#endif
	}
      else
	cout<<" -- wrong!\n";
    }      
    }
  abort();
}



