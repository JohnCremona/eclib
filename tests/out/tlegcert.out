Solving ax^2 + by^2 + cz^2 = 0
Using method 4

Enter coefficients a b c: 10000000019 10000000033 -10000010639
Enter certificate k1 k2 k3: Solution: (x:y:z) = (16081092938:20311652644:-25906834446)
 --OK
Enter coefficients a b c: 0 0 0
